ssl_test.go

```diff
@@ -0,0 +1,466 @@
+	}
+
+	return openssl.Listen(network, address, ctx.(*openssl.Ctx))
+}
+
+func serverSslAccept(l net.Listener) (<-chan string, <-chan error) {
+	message := make(chan string, 1)
+	errors := make(chan error, 1)
+
+	go func() {
+		conn, err := l.Accept()
+		if err != nil {
+			errors <- err
+		} else {
+			bytes, err := ioutil.ReadAll(conn)
+			if err != nil {
+				errors <- err
+			} else {
+				message <- string(bytes)
+			}
+			conn.Close()
+		}
+
+		close(message)
+		close(errors)
+	}()
+
+	return message, errors
+}
+
+func serverSslRecv(msgs <-chan string, errs <-chan error) (string, error) {
+	return <-msgs, <-errs
+}
+
+func clientSsl(network, address string, opts SslOpts) (net.Conn, error) {
+	timeout := 5 * time.Second
+	return SslDialTimeout(network, address, timeout, opts)
+}
+
+func createClientServerSsl(t testing.TB, serverOpts,
+	clientOpts SslOpts) (net.Listener, net.Conn, error, <-chan string, <-chan error) {
+	t.Helper()
+
+	l, err := serverSsl("tcp", sslHost+":0", serverOpts)
+	if err != nil {
+		t.Fatalf("Unable to create server, error %q", err.Error())
+	}
+
+	msgs, errs := serverSslAccept(l)
+
+	port := l.Addr().(*net.TCPAddr).Port
+	c, err := clientSsl("tcp", sslHost+":"+strconv.Itoa(port), clientOpts)
+
+	return l, c, err, msgs, errs
+}
+
+func createClientServerSslOk(t testing.TB, serverOpts,
+	clientOpts SslOpts) (net.Listener, net.Conn, <-chan string, <-chan error) {
+	t.Helper()
+
+	l, c, err, msgs, errs := createClientServerSsl(t, serverOpts, clientOpts)
+	if err != nil {
+		t.Fatalf("Unable to create client, error %q", err.Error())
+	}
+
+	return l, c, msgs, errs
+}
+
+func serverTnt(serverOpts, clientOpts SslOpts) (test_helpers.TarantoolInstance, error) {
+	listen := tntHost + "?transport=ssl&"
+
+	key := serverOpts.KeyFile
+	if key != "" {
+		listen += fmt.Sprintf("ssl_key_file=%s&", key)
+	}
+
+	cert := serverOpts.CertFile
+	if cert != "" {
+		listen += fmt.Sprintf("ssl_cert_file=%s&", cert)
+	}
+
+	ca := serverOpts.CaFile
+	if ca != "" {
+		listen += fmt.Sprintf("ssl_ca_file=%s&", ca)
+	}
+
+	ciphers := serverOpts.Ciphers
+	if ciphers != "" {
+		listen += fmt.Sprintf("ssl_ciphers=%s&", ciphers)
+	}
+
+	listen = listen[:len(listen)-1]
+
+	return test_helpers.StartTarantool(test_helpers.StartOpts{
+		InitScript:    "config.lua",
+		Listen:        listen,
+		SslCertsDir:   "testdata",
+		PingServer:    tntHost,
+		PingTransport: "ssl",
+		PingSsl:       clientOpts,
+		WorkDir:       "work_dir_ssl",
+		User:          "test",
+		Pass:          "test",
+		WaitStart:     100 * time.Millisecond,
+		ConnectRetry:  3,
+		RetryTimeout:  500 * time.Millisecond,
+	})
+}
+
+func serverTntStop(inst test_helpers.TarantoolInstance) {
+	test_helpers.StopTarantoolWithCleanup(inst)
+}
+
+func assertConnectionSslFail(t testing.TB, serverOpts, clientOpts SslOpts) {
+	t.Helper()
+
+	l, c, err, _, _ := createClientServerSsl(t, serverOpts, clientOpts)
+	if err == nil {
+		t.Fatalf("An unexpected connection to the server.")
+		c.Close()
```

Comment:
Code below `Fatalf` is not executed. Is it what we really want?

Reply:
Yes, you are right. Thank you, updated.